src/mapeo-manager.js

```diff
@@ -0,0 +1,148 @@
+import { randomBytes } from 'crypto'
+import path from 'path'
+import { KeyManager } from '@mapeo/crypto'
+import Database from 'better-sqlite3'
+import { eq } from 'drizzle-orm'
+import { drizzle } from 'drizzle-orm/better-sqlite3'
+import { migrate } from 'drizzle-orm/better-sqlite3/migrator'
+import { IndexWriter } from './index-writer/index.js'
+import { MapeoProject } from './mapeo-project.js'
+import { projectKeysTable, projectTable } from './schema/client.js'
+import { ProjectKeys } from './generated/keys.js'
+
+const CLIENT_SQLITE_FILE_NAME = 'client.db'
+
+export class MapeoManager {
+  #keyManager
+  #projectSettingsIndexWriter
+  #storagePath
+  #db
+  /** @type {Map<string, MapeoProject>} */
+  #activeProjects
+
+  /**
+   * @param {Object} [opts]
+   * @param {string} [opts.storagePath] Folder for all data storage (hypercores and sqlite db). Folder must exist. If not defined, everything is stored in-memory
+   */
+  constructor({ storagePath } = {}) {
+    const dbPath =
+      storagePath !== undefined
+        ? path.join(storagePath, CLIENT_SQLITE_FILE_NAME)
+        : ':memory:'
+
+    const sqlite = new Database(dbPath)
+    this.#db = drizzle(sqlite)
+    migrate(this.#db, { migrationsFolder: './drizzle/client' })
+
+    this.#keyManager = new KeyManager(KeyManager.generateRootKey())
+    this.#projectSettingsIndexWriter = new IndexWriter({
+      tables: [projectTable],
+      sqlite,
+    })
+    this.#storagePath = storagePath
+    this.#activeProjects = new Map()
+  }
+
+  /**
+   * Create a new project.
+   * @param {import('type-fest').Simplify<Partial<Pick<import("@mapeo/schema").ProjectValue, 'name'>>>} [settings]
+   * @returns {Promise<string>}
+   */
+  async createProject(settings = {}) {
+    // 1. Create project keypair
+    const projectKeypair = KeyManager.generateProjectKeypair()
+
+    // 2. Create namespace encryption keys
+    /** @type {Record<import('./core-manager/core-index.js').Namespace, Buffer>} */
+    const encryptionKeys = {
+      auth: randomBytes(32),
+      blob: randomBytes(32),
+      blobIndex: randomBytes(32),
+      config: randomBytes(32),
+      data: randomBytes(32),
+    }
+
+    // 3. Save keys to client db in projectKeys table
+    /** @type {ProjectKeys} */
+    const keys = {
+      projectKey: projectKeypair.publicKey,
+      projectSecretKey: projectKeypair.secretKey,
+      encryptionKeys,
+    }
+
+    // TODO: Update to use @mapeo/crypto when ready (https://github.com/digidem/mapeo-core-next/issues/171)
+    const projectId = projectKeypair.publicKey.toString('hex')
+
+    this.#db
+      .insert(projectKeysTable)
+      .values({
+        projectId,
+        keysCipher: this.#keyManager.encryptLocalMessage(
+          // TODO: Type error should go away once https://github.com/digidem/mapeo-crypto/issues/14 is resolved
+          // @ts-expect-error
+          ProjectKeys.encode(keys).finish(),
```

Comment:
```suggestion
          Buffer.from(ProjectKeys.encode(keys).finish().buffer),
```
see https://github.com/digidem/mapeo-crypto/issues/14. Using the `.buffer` [avoids making a copy](https://nodejs.org/api/buffer.html#buffer_static_method_buffer_from_arraybuffer_byteoffset_length)

Reply:
using `.buffer` leaves me with this 

![image](https://github.com/digidem/mapeo-core-next/assets/18542095/53a55b4f-a1f5-4999-a31c-bee2cfa42497)

Reply:
or am i supposed to still do `Buffer.from` here? i.e. `Buffer.from(ProjectKeys.encode(keys).finish().buffer)`

Reply:
Yes, I think if I understand the question, use as my suggested commit above (haven't checked it works)

Reply:
yeah forgot to call Buffer.from in my initial question - didn't actually read the suggestion when implementing the change (not a morning person 😄 )

Reply:
addressed via https://github.com/digidem/mapeo-core-next/pull/199/commits/a59a97a4e70a7cabb466920bac0810e0cb728730